plugins/connection/grpc.py

```diff
@@ -0,0 +1,292 @@
+          in place of network_os.
+    default: False
+    ini:
+      - section: grpc_connection
+        key: type
+    env:
+      - name: ANSIBLE_GRPC_CONNECTION_TYPE
+    vars:
+      - name: ansible_grpc_connection_type
+"""
+
+from importlib import import_module
+
+from ansible.errors import AnsibleConnectionFailure, AnsibleError
+from ansible.plugins.connection import NetworkConnectionBase
+
+try:
+    from grpc import insecure_channel, secure_channel, ssl_channel_credentials
+    from grpc.beta import implementations
+
+    HAS_GRPC = True
+except ImportError:
+    HAS_GRPC = False
+
+try:
+    from google import protobuf  # noqa: F401
+
+    HAS_PROTOBUF = True
+except ImportError:
+    HAS_PROTOBUF = False
+
+
+class Connection(NetworkConnectionBase):
+    """GRPC connections"""
+
+    transport = "ansible.netcommon.grpc"
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+
+        super(Connection, self).__init__(
+            play_context, new_stdin, *args, **kwargs
+        )
+
+        grpc_type = self._network_os or self.get_option("grpc_type")
+        if grpc_type:
+            if not HAS_PROTOBUF:
+                raise AnsibleError(
+                    "protobuf is required to use the grpc connection type. Please run 'pip install protobuf'"
+                )
+            if self._network_os:
+                os_split = self._network_os.split(".")
+                grpc_type = os_split[0] + "." + os_split[1] + ".grpc"
+            else:
+                os_split = grpc_type.split(".")
+                self._network_os = (
+                    os_split[0] + "." + os_split[1] + "." + os_split[1]
+                )
+            cref = dict(zip(["corg", "cname", "plugin"], grpc_type.split(".")))
+            grpclib = "ansible_collections.{corg}.{cname}.plugins.sub_plugins.grpc.{plugin}".format(
+                **cref
+            )
+            grpccls = getattr(import_module(grpclib), "Grpc")
+            grpc_obj = grpccls(self)
+
+            if grpc_obj:
+                self._sub_plugin = {
+                    "type": "grpc",
+                    "name": grpc_type,
+                    "obj": grpc_obj,
+                }
+                self.queue_message(
+                    "log", "loaded gRPC plugin for type %s" % grpc_type
+                )
+                self.queue_message("log", "grpc type is set to %s" % grpc_type)
+            else:
+                raise AnsibleConnectionFailure(
+                    "unable to load API plugin for network_os %s"
+                    % self._network_os
+                )
+        else:
+            self._sub_plugin["type"] = "external"
+            self.queue_message(
+                "warning",
+                "Unable to automatically determine gRPC implementation type."
+                " Please manually configure ansible_network_os value or grpc_type configuration for this host",
+            )
```

Comment:
Why isn't this a failure? `type: external` is for when the object in `obj` can't be loaded by PluginLoader, but that's not really the case here.

Reply:
Agreed. Fixing it.